signing_service/src/signing_service.py

```diff
@@ -0,0 +1,151 @@
+                # Try to close opened socket.
+                self.socket.close()
+                # If there was an unrecognized exception, log it and report to Sentry.
+                logger.error(f'Unrecognized exception occurred: {exception}')
+                crash_logger.error(str(exception))
+                raise
+
+    def _connect(self) -> None:
+        """ Creates socket and connects to given HOST and PORT. """
+        self.socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+
+        if self.current_reconnect_delay > 0:
+            logger.warning(f'Waiting {self.current_reconnect_delay} before connecting.')
+            sleep(self.current_reconnect_delay)
+
+        logger.warning(f'Connecting to {self.host}:{self.port}.')
+        self.socket.connect((self.host, self.port))
+        logger.warning(f'Connection established.')
+
+        # Reset delay on successful connection and set flag that connection is established.
+        self.current_reconnect_delay = 0
+        self._handle_connection()
+
+    def _handle_connection(self) -> None:
+        """ Inner loop that handles data exchange over socket. """
+        while True:
+            data = self.socket.recv(1024)
+            self.socket.send(data)
+
+    def _increase_delay(self) -> None:
+        """ Increase current delay if connection cannot be established. """
+        if self.current_reconnect_delay == 0:
+            self.current_reconnect_delay = self.reconnect_delay
+        elif self.current_reconnect_delay < SIGNING_SERVICE_MAXIMUM_RECONNECT_TIME:
+            self.current_reconnect_delay *= 2
+        elif self.current_reconnect_delay >= SIGNING_SERVICE_MAXIMUM_RECONNECT_TIME:
+            self.current_reconnect_delay = SIGNING_SERVICE_MAXIMUM_RECONNECT_TIME
+
+
+def _parse_arguments() -> argparse.Namespace:
+    parser = argparse.ArgumentParser()
+    parser.add_argument('concent_cluster_address', help='Address of a Concent cluster.')
+    parser.add_argument(
+        '--concent-cluster-port',
+        default=SIGNING_SERVICE_DEFAULT_PORT,
+        dest='concent_cluster_port',
+        type=int,
+        help=f'Port on which Concent cluster is listening (default: {SIGNING_SERVICE_DEFAULT_PORT}).'
+    )
+    parser.add_argument(
+        'reconnect_delay',
+        type=int,
+        help='Initial delay between reconnections, doubles after each unsuccessful attempt and is reset after success.',
+    )
+
+    return parser.parse_args()
+
+
+if __name__ == '__main__':
+    logging.config.fileConfig(os.path.join(os.path.dirname(os.path.abspath(__file__)), '..', 'logging.ini'))
+
+    # Parse required arguments.
+    args = _parse_arguments()
+
+    arg_host = args.concent_cluster_address
+    arg_port = args.concent_cluster_port
+    arg_reconnect_delay = args.reconnect_delay
+
+    if not 0 < arg_port < 65535:
+        exit('Port must be 0-65535.')
+
+    if arg_reconnect_delay < 0:
+        exit('reconnect_delay must be non-negative integer.')
```

Comment:
Maybe we should have `_validate_arguments()`?

Reply:
I am thinking if there is already enough code there to move it to separate function. Or maybe it should be a method of `SigningService` called in __init__ ?

Reply:
Maybe. But then it should raise exceptions and have its own, custom exception class. `exit()` should be only at the top level.

Reply:
:+1: done